soffice: Report keyboard-triggered font size change

michaelweghorn opened the pull request:
### Link to issue number:

Partially implements feature requests from #6915

### Summary of the issue:

commit 46a343685247ed7d1adfb737716590ec4be456a0
("soffice: Report keyboard-triggered formatting toggles in Writer (#16413)") implemented announcement of formatting changes
triggered by keyboard shortcuts for formatting attributes whose state is represented by toggle buttons in Writer's formatting toolbar.
Issue #6915 requests announcement of further text formatting attributes that are not represented by simple toggle buttons. This includes decreasing/increasing the font size, which can be done using the Ctrl+[ and Ctrl+] shortcuts when using an English (US) UI and keyboard layout.

### Description of user facing changes

When decreasing or increasing the font size in LibreOffice Writer using the corresponding keyboard shortcuts, NVDA announces the new font size.

### Description of development approach

Extend the solution from the above-mentioned
commit 46a343685247ed7d1adfb737716590ec4be456a0
to not only cover toggle buttons, but also
UI controls implementing the IAccessibleText
interface and handle the gestures/keyboard shortcuts for changing the font size:

* Rename methods and variables introduced earlier to not have button-specific names.
* Extract some logic to helper methods/classes to avoid duplication in `SymphonyButton` and the newly introduced `SymphonyText` logic.
* Add Ctrl+[ and Ctrl+] to the list of keyboard gestures to handle.
* Add `SymphonyText.event_valueChange` override that announces the new value. This gets triggered when the value of the "Font Size" editable combobox in Writer's formatting toolbar changes after using the keyboard shortcut, and results in the new value being announced, e.g. "14 pt". This is comparable to the handling in `SymphonyButton.event_stateChange` introduced in the earlier above-mentioned commit.
* Increase the timeout for announcement of events from 0.15 to 2 seconds, as 0.15 seconds wasn't always sufficient when testing the new feature.

### Testing strategy:

1. start NVDA
2. start LibreOffice Writer (with English UI and keyboard layout)
3. Press Ctrl+[ to decrease the font size
4. Verify that NVDA announces the new font size, e.g. "10 pt" when using the default document template
5. Press Ctrl+] to increase the font size
4. Verify that NVDA announces the new font size, e.g. "12 pt" when using the default document template
5. repeat the tests mentioned in commit 46a343685247ed7d1adfb737716590ec4be456a0 to verify the toggle button case still works as expected

### Known issues with pull request:

None

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.


<!-- Please keep the following -->
@coderabbitai summary

michaelweghorn committed:
soffice: Report keyboard-triggered font size change

 ### Link to issue number:

Partially implements feature requests from #6915

 ### Summary of the issue:

commit 46a343685247ed7d1adfb737716590ec4be456a0
("soffice: Report keyboard-triggered formatting toggles in Writer (#16413)")
implemented announcement of formatting changes
triggered by keyboard shortcuts for formatting attributes
whose state is represented by toggle buttons in Writer's
formatting toolbar.
Issue #6915 requests announcement of further text formatting
attributes that are not represented by simple toggle buttons.
This includes decreasing/increasing the font size, which
can be done using the Ctrl+[ and Ctrl+] shortcuts when
using an English (US) UI and keyboard layout.

 ### Description of user facing changes

When decreasing or increasing the font size in LibreOffice
Writer using the corresponding keyboard shortcuts, NVDA
announces the new font size.

 ### Description of development approach

Extend the solution from the above-mentioned
commit 46a343685247ed7d1adfb737716590ec4be456a0
to not only cover toggle buttons, but also
UI controls implementing the IAccessibleText
interface and handle the gestures/keyboard shortcuts
for changing the font size:

* Rename methods and variables introduced earlier
  to not have button-specific names.
* Extract some logic to helper methods/classes to
  avoid duplication in `SymphonyButton` and the
  newly introduced `SymphonyText` logic.
* Add Ctrl+[ and Ctrl+] to the list of keyboard gestures
  to handle.
* Add `SymphonyText.event_valueChange` override
  that announces the new value. This gets triggered
  when the value of the "Font Size" editable combobox
  in Writer's formatting toolbar changes after
  using the keyboard shortcut, and results in the
  new value being announced, e.g. "14 pt".
  This is comparable to the handling in
  `SymphonyButton.event_stateChange`
  introduced in the earlier above-mentioned commit.
* Increase the timeout for announcement of events
  from 0.15 to 2 seconds, as 0.15 seconds wasn't always
  sufficient when testing the new feature.

 ### Testing strategy:

1. start NVDA
2. start LibreOffice Writer (with English UI and
   keyboard layout)
3. Press Ctrl+[ to decrease the font size
4. Verify that NVDA announces the new font size,
   e.g. "10 pt" when using the default document template
5. Press Ctrl+] to increase the font size
4. Verify that NVDA announces the new font size,
   e.g. "12 pt" when using the default document template
5. repeat the tests mentioned in commit
   46a343685247ed7d1adfb737716590ec4be456a0 to verify
   the toggle button case still works as expected

 ### Known issues with pull request:

None

 ### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -431,10 +447,16 @@ class SymphonyDocument(CompoundDocument):
 
 	# variables used for handling announcements resulting from gestures
-	GESTURE_ANNOUNCEMENT_TIMEOUT = 0.15
-	announceToolbarButtonToggle = False
```
this appears to be an API breaking change. Can you please document it in the API breaking changes section of the change log.

michaelweghorn replied:
Done now. (I wasn't really aware that app modules are part of the public API as well.)

seanbudd replied:
thanks - they are, its rarely an issue but better be safe than sorry 😄

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -431,10 +447,16 @@ class SymphonyDocument(CompoundDocument):
 	lastFormattingGestureEventTime = 0
 
+	@staticmethod
+	def isFormattingChangeAnnouncementEnabled():
```
```suggestion
	def isFormattingChangeAnnouncementEnabled() -> bool:
```

seanbudd reviewed (commented):
Thanks @michaelweghorn - looks almost ready to go

michaelweghorn committed:
Add annotation for return type

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

michaelweghorn committed:
changelog: Mention API breaking changes, use correct issue number

While issue #16413 also requested announcement of formatting changes,
PR #17147 is for issue #6915.
